charms/k8s/src/charm.py

```diff
@@ -128,6 +145,36 @@ def _install_k8s_snap(self):
             channel = self.config["channel"]
             k8s_snap.ensure(SnapState.Latest, channel=channel)
 
+    @on_error(WaitingStatus("Waiting for Cluster token"), TypeError)
+    def _join_cluster(self):
+        """Retrieve the join token from secret databag and join the cluster."""
+        if self.unit.is_leader() or self._stored.joined:
+            return
+
+        status.add(ops.MaintenanceStatus("Joining cluster"))
+
+        if relation := self.model.get_relation("cluster"):
+            app_databag = relation.data.get(self.model.app, {})
+            secret_id = app_databag.get(self.unit.name, "")
+            secret = self.model.get_secret(id=secret_id)
+            content = secret.get_content()
+            token = content["token"]
+            cmd = f"k8s join-cluster {shlex.quote(token)}"
+            subprocess.check_call(shlex.split(cmd))
+            self._stored.joined = True
+
+    def _reconcile(self, _):
+        """Reconcile state change events."""
+        self._install_k8s_snap()
+        self._apply_snap_requirements()
+        if self.unit.is_leader():
+            self._bootstrap_k8s_snap()
```

Comment:
what if `_bootstrap_k8s_snap()` sets the `self._stored.joined=True`??  I don't suppose there's anyway to ask the snap itself if this node is part of the cluster?

Reply:
Yes, I have a better idea to avoid using StoredState, we can leverage the `is_cluster_bootstrapped check. Thanks for starting my train of thought @addyess 🚆